site/learn/expert-advice/security/breached-password-detection.md

```diff
@@ -0,0 +1,155 @@
+
+## Taking action when the breached password is found
+
+Alright, we've built this capability and now have found a user with a breached password. What do we need to do? The best choice depends on the level of harm unauthorized access could cause. 
+
+In increasing order of end user impact, the user management service could:
+
+* Update an internal system, but not notify the end user. This might be useful if you don't protect any data, but rather use authentication to throttle access or for reporting. If you are gradually rolling breached password detection out and expect user pushback when they are required to take action, you may want to get a baseline for how many compromised credentials are in your system currently.
+* Notify the user by sending them an email, or contacting them in some other fashion, to let them know about the compromised credential. I'd suggest politely asking them to change their password at that time.
+* Force them to reset their password. Most user management systems have a way to expire a password, so you can tie directly into that.
+* Lock the account and prohibit any use until the situation is rectified. This may include resetting the password or other steps such as investigating to confirm there was not any unauthorized access.
+
+Consider the privilege level of the account with the breached password. If the user is an administrator, you may want to take different actions than you would if they are a normal user.
+
+## External integration
+
+Depending on the breadth of your digital infrastructure, you may need to integrate with external systems when an account credential is compromised. Whether this happens in an entirely automated fashion or kicks off a manual process depends, again, on your security posture. Ensure the authentication system you are integrating with can fire off events notifying other interested systems.
+
+You also may want to be able to run reports or other analytics to determine if there are any patterns to the compromised passwords, or simply to know how many of your users have been affected. These reports may be provided in the auth system. An alternative is to use or build an analytics system to, again, subscribe to the published password breach events. 
+
+## Performance
+
+All security is a tradeoff; in this case, checking if password is breached is going to have an impact. The question is, how large is it? Performing a hash against a password isn't too computationally intensive. If there are network calls required, it may take a bit longer. It also depends on how many registration, password change and login attempts occur over a given timeframe. Only you can test in a real-world scenario and determine the performance impact on your system, but a back of the envelope calculation may help in deciding how many resources to spend. 
```

Comment:
Technically, computing hashes is supposed to be extremely computationally expensive. That's why the recommendation is to use something like Bcrypt with a load of 14. That way only 20 hashes per second can be generated on a beefy GPU. And technically, the hash isn't part of the breached check. 

You might want to reword this part a bit.